core/src/main/scala/kafka/log/AbstractIndex.scala

```diff
@@ -107,10 +128,14 @@ abstract class AbstractIndex[K, V](@volatile var file: File, val baseOffset: Lon
 
   protected val lock = new ReentrantLock
 
-  @volatile
-  protected var mmap: MappedByteBuffer = {
-    val newlyCreated = file.createNewFile()
-    val raf = if (writable) new RandomAccessFile(file, "rw") else new RandomAccessFile(file, "r")
+
+  @volatile protected var _mmap: MappedByteBuffer = null
+
+  @volatile var initialized = false
+
+  private def initializeMmapAndLength: Unit = {
+    val newlyCreated = _file.createNewFile()
+    val raf = if (writable) new RandomAccessFile(_file, "rw") else new RandomAccessFile(_file, "r")
```

Comment:
more DRY to say:

    val raf = new RandomAccessFile(_file, if (writable) "rw" else "r")

Reply:
Fixed.